tsdb/engine/tsm1/reader.go

```diff
@@ -1233,6 +1233,10 @@ func (d *indirectIndex) UnmarshalBinary(b []byte) error {
 			return fmt.Errorf("indirectIndex: not enough data for index entries count")
 		}
 		count := int32(binary.BigEndian.Uint16(b[i : i+indexCountSize]))
+
+		if count <= 0 {
+			return fmt.Errorf("invalid index %d; count must be greater than 0", count)
+		}
```

Comment:
There is no way for count to be less than 0. All possible values of a uint16 will fit in an int32 and since uint16 is only positive, count will never be less than 0.

`count` could be 0, though, and we definitely don't want to continue further in this case. It's possible `count == 0` is valid, but having logged is good for now.

Reply:
The fundamental solution for this case which I have considered was to add CRC for whole **_Index_** block as well, as it is done for other parts.  But the solution obviously will break backward compatibility. 
Are there any mechanism allowing to have support of different incompatible versions of *tsm file, with special hangling? 
Another thing I have taken account during implementation is the following check in  writer code _**tsdb/engine/tsm1/writer.go**_: 708 line , which skips _count == 0_  cases.

```
func (t *tsmWriter) WriteIndex() error {
	indexPos := t.n

	if t.index.KeyCount() == 0 {
		return ErrNoValues
	}
```

Reply:
It is unlikely we will introduce an incompatible TSM format at this point in the product's lifecycle.